File: data/verified_projects/commons-validator/src/test/org/apache/commons/validator/UrlTest.py
import pytest

from src.main.org.apache.commons.validator.UrlValidator import *
from src.test.org.apache.commons.validator.ResultPair import ResultPair
import unittest

class UrlTest(unittest.TestCase):

    __printStatus = False
    __printIndex = False

    testUrlScheme = [
        ResultPair("http://", True),
        ResultPair("ftp://", True),
        ResultPair("h3t://", True),
        ResultPair("3ht://", False),
        ResultPair("http:/", False),
        ResultPair("http:", False),
        ResultPair("http/", False),
        ResultPair("://", False),
        ResultPair("", True)
    ]
    testUrlAuthority = [
        ResultPair("www.google.com", True),
        ResultPair("go.com", True),
        ResultPair("go.au", True),
        ResultPair("0.0.0.0", True),
        ResultPair("255.255.255.255", True),
        ResultPair("256.256.256.256", False),
        ResultPair("255.com", True),
        ResultPair("1.2.3.4.5", False),
        ResultPair("1.2.3.4.", False),
        ResultPair("1.2.3", False),
        ResultPair(".1.2.3.4", False),
        ResultPair("go.a", False),
        ResultPair("go.a1a", True),
        ResultPair("go.1aa", False),
        ResultPair("aaa.", False),
        ResultPair(".aaa", False),
        ResultPair("aaa", False),
        ResultPair("", False)
    ]
    testUrlPort = [
        ResultPair(":80", True),
        ResultPair(":65535", True),
        ResultPair(":0", True),
        ResultPair("", True),
        ResultPair(":-1", False),
        ResultPair(":65636", True),
        ResultPair(":65a", False)
    ]
    testPath = [
        ResultPair("/test1", True),
        ResultPair("/t123", True),
        ResultPair("/$23", True),
        ResultPair("/..", False),
        ResultPair("/../", False),
        ResultPair("/test1/", True),
        ResultPair("", True),
        ResultPair("/test1/file", True),
        ResultPair("/..//file", False),
        ResultPair("/test1//file", False)
    ]
    testUrlPathOptions = [
        ResultPair("/test1", True),
        ResultPair("/t123", True),
        ResultPair("/$23", True),
        ResultPair("/..", False),
        ResultPair("/../", False),
        ResultPair("/test1/", True),
        ResultPair("/#", False),
        ResultPair("", True),
        ResultPair("/test1/file", True),
        ResultPair("/t123/file", True),
        ResultPair("/$23/file", True),
        ResultPair("/../file", False),
        ResultPair("/..//file", False),
        ResultPair("/test1//file", True),
        ResultPair("/#/file", False)
    ]

    testUrlQuery = [
        ResultPair("?action=view", True),
        ResultPair("?action=edit&mode=up", True),
        ResultPair("", True)
    ]

    testUrlParts = [
        testUrlScheme,
        testUrlAuthority,
        testUrlPort,
        testPath,
        testUrlQuery
    ]
    testUrlPartsOptions = [
        testUrlScheme,
        testUrlAuthority,
        testUrlPort,
        testUrlPathOptions,
        testUrlQuery
    ]
    testPartsIndex = [0, 0, 0, 0, 0]

    testScheme = [
        ResultPair("http", True),
        ResultPair("ftp", False),
        ResultPair("httpd", False),
        ResultPair("telnet", False)
    ]
    

    def setUp(self) -> None:
        super().setUp()
        for index in range(len(self.testPartsIndex) - 1):
            self.testPartsIndex[index] = 0

    
    @pytest.mark.test
    def testIsValid0(self) -> None:
        self.checkTestIsValid1(self.testUrlParts, UrlValidator.ALLOW_ALL_SCHEMES)
        self.setUp()
        options = UrlValidator.ALLOW_2_SLASHES +\
            UrlValidator.ALLOW_ALL_SCHEMES +\
            UrlValidator.NO_FRAGMENTS
        self.checkTestIsValid1(self.testUrlPartsOptions, options)
    

    @pytest.mark.test
    def testIsValidScheme(self) -> None:
        if self.__printStatus:
            print("\n testIsValidScheme() ")
        schemes = ["http", "gopher"]
        urlVal = UrlValidator(schemes, 0)
        for sIndex in range(len(self.testScheme)):
            testPair = self.testScheme[sIndex]
            result = urlVal.isValidScheme(testPair.item)
            self.assertEqual(
                testPair.valid,
                result,
                testPair.item
            )
            if self.__printStatus:
                if result == testPair.valid:
                    print('.', end='')
                else:
                    print('X', end='')
        if self.__printStatus:
            print()
    

    @pytest.mark.test
    def testValidator202(self) -> None:
        schemes = ["http", "https"]
        urlValidator = UrlValidator(schemes, UrlValidator.NO_FRAGMENTS)
        urlValidator.isValid(
            "http://www.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks." +\
                "comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks." +\
                "comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks." +\
                "comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks." +\
                "comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks." +\
                "comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks." +\
                "comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks." +\
                "comwww.logoworks.comwww.log"
        )


    @pytest.mark.test
    def testValidator204(self) -> None:
        schemes = ["http", "https"]
        urlValidator = UrlValidator.UrlValidator2(schemes)
        self.assertTrue(
            urlValidator.isValid(
                "http://tech.yahoo.com/rc/desktops/102;_ylt=Ao8yevQHlZ4On0O3ZJGXLEQFLZA5"
            )
        )
    

    @pytest.mark.test
    def testValidateUrl(self) -> None:
        self.assertTrue(True)
    

    @staticmethod
    def main(argv) -> None:
        fct = UrlTest()
        fct.setUp()
        fct.testIsValid0()
        fct.testIsValidScheme()


    def checkTestIsValid1(self, testObjects, options) -> None:
        urlVal = UrlValidator.UrlValidator1(None, None, options)
        self.assertTrue(urlVal.isValid("http://www.google.com"))
        self.assertTrue(urlVal.isValid("http://www.google.com/"))
        statusPerLine = 60
        printed = 0
        if self.__printIndex:
            statusPerLine = 6
        while True:
            testBuffer = []
            expected = True
            for testPartsIndexIndex in range(len(self.testPartsIndex)):
                index = self.testPartsIndex[testPartsIndexIndex]
                part = testObjects[testPartsIndexIndex]
                testBuffer.append(part[index].item)
                expected = (expected and part[index].valid)
            url = ''.join(testBuffer)
            result = urlVal.isValid(url)
            self.assertEqual(
                expected,
                result,
                url
            )
            if self.__printStatus:
                if self.__printIndex:
                    print(self.__testPartsIndextoString(), end='')
                else:
                    if result == expected:
                        print('.', end='')
                    else:
                        print('X', end='')
                printed += 1
                if printed == statusPerLine:
                    print()
                    printed = 0
            if not UrlTest.incrementTestPartsIndex(self.testPartsIndex, testObjects):
                break
        if self.__printStatus:
            print()
    

    @staticmethod
    def incrementTestPartsIndex(testPartsIndex, testParts) -> bool:
        carry = True
        maxIndex = True
        for testPartsIndexIndex in range(len(testPartsIndex) - 1, -1, -1):
            index = testPartsIndex[testPartsIndexIndex]
            part = testParts[testPartsIndexIndex]
            if carry:
                if index < len(part) - 1:
                    index += 1
                    testPartsIndex[testPartsIndexIndex] = index
                    carry = False
                else:
                    testPartsIndex[testPartsIndexIndex] = 0
                    carry = True
            maxIndex = (maxIndex and (index == (len(part) - 1)))
        
        return not maxIndex


    def __testPartsIndextoString(self) -> str:
        carryMsg = ['{']
        for testPartsIndexIndex in range(len(self.testPartsIndex)):
            carryMsg.append(str(self.testPartsIndex[testPartsIndexIndex]))
            if testPartsIndexIndex < len(self.testPartsIndex) - 1:
                carryMsg.append(',')
            else:
                carryMsg.append('}')
        return ''.join(carryMsg)
        

File: java_projects/cleaned_final_projects_evosuite/commons-validator/src/test/java/org/apache/commons/validator/UrlTest.java

/*
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

package org.apache.commons.validator;

import junit.framework.TestCase;

/**
 * Performs Validation Test for url validations.
 *
 * @version $Revision$
 * @deprecated to be removed when org.apache.commons.validator.UrlValidator is removed
 */
@Deprecated
public class UrlTest extends TestCase {

    private final boolean printStatus = false;
    private final boolean printIndex =
            false; // print index that indicates current scheme,host,port,path, query test were

    public UrlTest(String testName) {
        super(testName);
    }

    @Override
    protected void setUp() {
        for (int index = 0; index < testPartsIndex.length - 1; index++) {
            testPartsIndex[index] = 0;
        }
    }

    public void testIsValid0() {
        testIsValid1(testUrlParts, UrlValidator.ALLOW_ALL_SCHEMES);
        setUp();
        int options =
                UrlValidator.ALLOW_2_SLASHES
                        + UrlValidator.ALLOW_ALL_SCHEMES
                        + UrlValidator.NO_FRAGMENTS;

        testIsValid1(testUrlPartsOptions, options);
    }

    public void testIsValidScheme() {
        if (printStatus) {
            System.out.print("\n testIsValidScheme() ");
        }
        String[] schemes = {"http", "gopher"};
        UrlValidator urlVal = new UrlValidator(schemes, 0);
        for (int sIndex = 0; sIndex < testScheme.length; sIndex++) {
            ResultPair testPair = testScheme[sIndex];
            boolean result = urlVal.isValidScheme(testPair.item);
            assertEquals(testPair.item, testPair.valid, result);
            if (printStatus) {
                if (result == testPair.valid) {
                    System.out.print('.');
                } else {
                    System.out.print('X');
                }
            }
        }
        if (printStatus) {
            System.out.println();
        }
    }

    /**
     * Create set of tests by taking the testUrlXXX arrays and running through all possible
     * permutations of their combinations.
     *
     * @param testObjects Used to create a url.
     */
    public void testIsValid1(Object[] testObjects, int options) {
        UrlValidator urlVal = new UrlValidator(null, options);
        assertTrue(urlVal.isValid("http://www.google.com"));
        assertTrue(urlVal.isValid("http://www.google.com/"));
        int statusPerLine = 60;
        int printed = 0;
        if (printIndex) {
            statusPerLine = 6;
        }
        do {
            StringBuilder testBuffer = new StringBuilder();
            boolean expected = true;
            for (int testPartsIndexIndex = 0;
                    testPartsIndexIndex < testPartsIndex.length;
                    ++testPartsIndexIndex) {
                int index = testPartsIndex[testPartsIndexIndex];
                ResultPair[] part = (ResultPair[]) testObjects[testPartsIndexIndex];
                testBuffer.append(part[index].item);
                expected &= part[index].valid;
            }
            String url = testBuffer.toString();
            boolean result = urlVal.isValid(url);
            assertEquals(url, expected, result);
            if (printStatus) {
                if (printIndex) {
                    System.out.print(testPartsIndextoString());
                } else {
                    if (result == expected) {
                        System.out.print('.');
                    } else {
                        System.out.print('X');
                    }
                }
                printed++;
                if (printed == statusPerLine) {
                    System.out.println();
                    printed = 0;
                }
            }
        } while (incrementTestPartsIndex(testPartsIndex, testObjects));
        if (printStatus) {
            System.out.println();
        }
    }

    public void testValidator202() {
        String[] schemes = {"http", "https"};
        UrlValidator urlValidator = new UrlValidator(schemes, UrlValidator.NO_FRAGMENTS);
        urlValidator.isValid(
                "http://www.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.log");
    }

    public void testValidator204() {
        String[] schemes = {"http", "https"};
        UrlValidator urlValidator = UrlValidator.UrlValidator2(schemes);
        assertTrue(
                urlValidator.isValid(
                        "http://tech.yahoo.com/rc/desktops/102;_ylt=Ao8yevQHlZ4On0O3ZJGXLEQFLZA5"));
    }

    static boolean incrementTestPartsIndex(int[] testPartsIndex, Object[] testParts) {
        boolean carry = true; // add 1 to lowest order part.
        boolean maxIndex = true;
        for (int testPartsIndexIndex = testPartsIndex.length - 1;
                testPartsIndexIndex >= 0;
                --testPartsIndexIndex) {
            int index = testPartsIndex[testPartsIndexIndex];
            ResultPair[] part = (ResultPair[]) testParts[testPartsIndexIndex];
            if (carry) {
                if (index < part.length - 1) {
                    index++;
                    testPartsIndex[testPartsIndexIndex] = index;
                    carry = false;
                } else {
                    testPartsIndex[testPartsIndexIndex] = 0;
                    carry = true;
                }
            }
            maxIndex &= (index == (part.length - 1));
        }

        return (!maxIndex);
    }

    private String testPartsIndextoString() {
        StringBuilder carryMsg = new StringBuilder("{");
        for (int testPartsIndexIndex = 0;
                testPartsIndexIndex < testPartsIndex.length;
                ++testPartsIndexIndex) {
            carryMsg.append(testPartsIndex[testPartsIndexIndex]);
            if (testPartsIndexIndex < testPartsIndex.length - 1) {
                carryMsg.append(',');
            } else {
                carryMsg.append('}');
            }
        }
        return carryMsg.toString();
    }

    public void testValidateUrl() {
        assertTrue(true);
    }

    /**
     * Only used to debug the unit tests.
     *
     * @param argv
     */
    public static void main(String[] argv) {

        UrlTest fct = new UrlTest("url test");
        fct.setUp();
        fct.testIsValid0();
        fct.testIsValidScheme();
    }

    /**
     * The data given below approximates the 4 parts of a URL <scheme>://<authority><path>?<query>
     * except that the port number is broken out of authority to increase the number of
     * permutations. A complete URL is composed of a scheme+authority+port+path+query, all of which
     * must be individually valid for the entire URL to be considered valid.
     */
    ResultPair[] testUrlScheme = {
        new ResultPair("http://", true),
        new ResultPair("ftp://", true),
        new ResultPair("h3t://", true),
        new ResultPair("3ht://", false),
        new ResultPair("http:/", false),
        new ResultPair("http:", false),
        new ResultPair("http/", false),
        new ResultPair("://", false),
        new ResultPair("", true)
    };

    ResultPair[] testUrlAuthority = {
        new ResultPair("www.google.com", true),
        new ResultPair("go.com", true),
        new ResultPair("go.au", true),
        new ResultPair("0.0.0.0", true),
        new ResultPair("255.255.255.255", true),
        new ResultPair("256.256.256.256", false),
        new ResultPair("255.com", true),
        new ResultPair("1.2.3.4.5", false),
        new ResultPair("1.2.3.4.", false),
        new ResultPair("1.2.3", false),
        new ResultPair(".1.2.3.4", false),
        new ResultPair("go.a", false),
        new ResultPair("go.a1a", true),
        new ResultPair("go.1aa", false),
        new ResultPair("aaa.", false),
        new ResultPair(".aaa", false),
        new ResultPair("aaa", false),
        new ResultPair("", false)
    };
    ResultPair[] testUrlPort = {
        new ResultPair(":80", true),
        new ResultPair(":65535", true),
        new ResultPair(":0", true),
        new ResultPair("", true),
        new ResultPair(":-1", false),
        new ResultPair(":65636", true),
        new ResultPair(":65a", false)
    };
    ResultPair[] testPath = {
        new ResultPair("/test1", true),
        new ResultPair("/t123", true),
        new ResultPair("/$23", true),
        new ResultPair("/..", false),
        new ResultPair("/../", false),
        new ResultPair("/test1/", true),
        new ResultPair("", true),
        new ResultPair("/test1/file", true),
        new ResultPair("/..//file", false),
        new ResultPair("/test1//file", false)
    };
    ResultPair[] testUrlPathOptions = {
        new ResultPair("/test1", true),
        new ResultPair("/t123", true),
        new ResultPair("/$23", true),
        new ResultPair("/..", false),
        new ResultPair("/../", false),
        new ResultPair("/test1/", true),
        new ResultPair("/#", false),
        new ResultPair("", true),
        new ResultPair("/test1/file", true),
        new ResultPair("/t123/file", true),
        new ResultPair("/$23/file", true),
        new ResultPair("/../file", false),
        new ResultPair("/..//file", false),
        new ResultPair("/test1//file", true),
        new ResultPair("/#/file", false)
    };

    ResultPair[] testUrlQuery = {
        new ResultPair("?action=view", true),
        new ResultPair("?action=edit&mode=up", true),
        new ResultPair("", true)
    };

    Object[] testUrlParts = {testUrlScheme, testUrlAuthority, testUrlPort, testPath, testUrlQuery};
    Object[] testUrlPartsOptions = {
        testUrlScheme, testUrlAuthority, testUrlPort, testUrlPathOptions, testUrlQuery
    };
    int[] testPartsIndex = {0, 0, 0, 0, 0};

    ResultPair[] testScheme = {
        new ResultPair("http", true),
        new ResultPair("ftp", false),
        new ResultPair("httpd", false),
        new ResultPair("telnet", false)
    };
}
